workflow/torch/datastream.py

```diff
@@ -104,7 +101,33 @@ def merge(datastreams_and_ns):
         )
 
         return Datastream(
-            concat_dataset,
+            Dataset.concat(datasets),
+            sampler,
+        )
+
+    @staticmethod
+    def _combine_samplers(samplers, map_index):
+        create_sampler = starcompose(
+            partial(map, partial(repeat_map_chain, iter)),
+            tuple,
+            zip,
+            partial(map, map_index),
+        )
+        return create_sampler(samplers)
+
+    @staticmethod
+    def combine(datastreams):
+        datasets = [datastream.dataset for datastream in datastreams]
+        samplers = [datastream.sampler for datastream in datastreams]
+        sampler = BoundlessSampler(
+            lambda: Datastream._combine_samplers(
+                samplers,
+                Dataset.create_to_combine_mapping(datasets),
+            ),
+            length=max(map(len, datasets)) * len(datasets),
```

Comment:
Is this correct? Should we remove `len(datasets)`?

Reply:
Good point, I agree that that definition doesn't make sense. The length of this dataset is actually defined by the sampling algorithms so it's kind of difficult to define a length here. For example, if all samplers are sequential and all datasets are the same length, the length of the dataset is `len(datasets[0])`. If they are all random samplers it is `np.prod(list(map(len, datasets)))`. If they are sequential and different lengths it gets tricky (nr of possible combinations now depend on prime factorizations of the lengths of the datasets...).

I think the most common use case is random sampling any combination of datapoints. Do we run into any troubles by defining it as `np.prod(list(map(len, datasets)))`?

Reply:
This is very similar to `Datastream.merge` right? But instead of interleaving the examples we are keeping them as tuples. Therefore I think it makes sense to use the same convention and set `length=max(map(len, datasets)),`. What do you think?

Reply:
That makes sense, I'll update to that :)